hazelcast/src/main/java/com/hazelcast/internal/probing/ProbingCycle.java

```diff
@@ -0,0 +1,222 @@
+     * dependent.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * 
+     * @param instance a obj with fields or methods annotated with {@link Probe}
+     */
+    void probe(Object instance);
+
+    /**
+     * Default method to call {@link #probe(Object)} for an array of instances. This
+     * is mostly useful for types that implement {@link Tagging} and thereby provide
+     * their own context.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * Instances do not affect each others context.
+     * 
+     * @param instances may be null (no effect)
+     */
+    void probe(Object[] instances);
+
+    /**
+     * Similar to {@link #probe(Object)} just that all names become
+     * {@code <prefix>.<name>} instead of just {@code <name>}.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * 
+     * @param prefix prefix to use for all names (provided without dot)
+     * @param instance a obj with fields or methods annotated with {@link Probe}
+     */
+    void probe(CharSequence prefix, Object instance);
+
+    /**
+     * Similar to {@link #probe(Object)} just that probed methods are not identified
+     * using annotations but given as a list. All methods use the specified level.
+     * No fields will be probed.
+     * 
+     * This can be used to probe beans that cannot be annotated. To prefix this
+     * {@link Tags#prefix(CharSequence)} can be used. Remember though that the
+     * prefix will remain in the context.
+     * 
+     * @param level the level to use for all probes created
+     * @param instance the obj whose methods to probe.
+     * @param methods the names of the methods to probe. Methods that do not exist
+     *        or have the wrong number of arguments (not none) are ignored.
+     */
+    void probe(ProbeLevel level, Object instance, String... methods);
+
+    void probe(CharSequence name, long value);
+
+    void probe(CharSequence name, double value);
+
+    void probe(CharSequence name, boolean value);
+
+    void probe(ProbeLevel level, CharSequence name, long value);
+
+    void probe(ProbeLevel level, CharSequence name, double value);
+
+    void probe(ProbeLevel level, CharSequence name, boolean value);
+
+    /**
+     * Used to forward metrics received from the client. This is only different from
+     * {@link #probe(CharSequence, long)} except that it will not escape the name
+     * but un-escaping escaped backslashes.
+     * 
+     * @param name a properly escaped name (as written before)
+     * @param value value as written before (double already converted to long)
+     */
+    void probeForwarded(CharSequence name, long value);
+
+    /*
+     * Describing probing context to the cycle:
+     */
+
+    /**
+     * Resets the tags used by this cycle (to empty). This provides a convenient and
```

Comment:
Do we want to recycle Tags? Because we don't want to create litter if not needed.

Reply:
Resetting internally sets length of buffer to zero. No objects created.

Reply:
Great